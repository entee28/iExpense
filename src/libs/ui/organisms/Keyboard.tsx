import { faDeleteLeft } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome'
import React from 'react'
import { Insets, StyleSheet, ViewStyle } from 'react-native'
import { Box, Pressable, Text } from '../atoms'
import colors from '../colors'

type Props = ViewStyle & {
  onKeyPress: (key: string) => void
  integerOnly?: boolean
}

const KEYS = [
  ['1', '2', '3'],
  ['4', '5', '6'],
  ['7', '8', '9'],
  ['.', '0', 'del']
]

export const Keyboard = (props: Props) => {
  const { onKeyPress, integerOnly, ...viewStyle } = props

  return (
    <Box {...viewStyle}>
      {KEYS.map((row, index) => (
        <Box
          flexDirection="row"
          justifyContent="space-around"
          key={`key-row-${index}`}
          marginTop={index > 0 ? 12 : 0}>
          {row.map(key => (
            <Pressable
              key={`key-${key}`}
              hitSlop={HIT_SLOP}
              onPress={
                key === '.' && integerOnly ? undefined : () => onKeyPress(key)
              }
              style={styles.key}>
              {key !== 'del' ? (
                <Text fontSize={32} lineHeight={38} bold textAlign="center">
                  {key === '.' && integerOnly ? '' : key}
                </Text>
              ) : (
                <FontAwesomeIcon
                  icon={faDeleteLeft}
                  size={24}
                  color={colors.mono100}
                />
              )}
            </Pressable>
          ))}
        </Box>
      ))}
    </Box>
  )
}

const HIT_SLOP: Insets = {
  top: 10,
  right: 10,
  bottom: 10,
  left: 10
}

const styles = StyleSheet.create({
  key: {
    width: 48,
    height: 48,
    justifyContent: 'center',
    alignItems: 'center'
  }
})
