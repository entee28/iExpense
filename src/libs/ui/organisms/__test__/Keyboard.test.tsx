import { fireEvent, render } from '@testing-library/react-native'
import { Keyboard } from '../Keyboard'

it('should handle key press correctly', () => {
  let string = ''
  const handleKeyPress = (key: string) => {
    string += key
  }

  const keyboard = render(<Keyboard onKeyPress={handleKeyPress} />)

  fireEvent.press(keyboard.getByText('1'))
  fireEvent.press(keyboard.getByText('2'))
  fireEvent.press(keyboard.getByText('3'))
  fireEvent.press(keyboard.getByText('4'))
  fireEvent.press(keyboard.getByText('5'))
  fireEvent.press(keyboard.getByText('6'))
  fireEvent.press(keyboard.getByText('7'))
  fireEvent.press(keyboard.getByText('8'))
  fireEvent.press(keyboard.getByText('9'))
  fireEvent.press(keyboard.getByText('.'))
  fireEvent.press(keyboard.getByText('0'))

  expect(string).toEqual('123456789.0')
})

it('should match snapshot', () => {
  const keyboard = render(<Keyboard onKeyPress={() => {}} />).toJSON()
  expect(keyboard).toMatchSnapshot()
})
