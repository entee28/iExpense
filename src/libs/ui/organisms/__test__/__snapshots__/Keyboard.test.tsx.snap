// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  style={{}}
>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginTop": 0,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1A1A1A",
              "fontFamily": "Nunito-Bold",
              "fontSize": 32,
              "lineHeight": 38,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        1
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1A1A1A",
              "fontFamily": "Nunito-Bold",
              "fontSize": 32,
              "lineHeight": 38,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        2
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1A1A1A",
              "fontFamily": "Nunito-Bold",
              "fontSize": 32,
              "lineHeight": 38,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        3
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginTop": 12,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1A1A1A",
              "fontFamily": "Nunito-Bold",
              "fontSize": 32,
              "lineHeight": 38,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        4
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1A1A1A",
              "fontFamily": "Nunito-Bold",
              "fontSize": 32,
              "lineHeight": 38,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        5
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1A1A1A",
              "fontFamily": "Nunito-Bold",
              "fontSize": 32,
              "lineHeight": 38,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        6
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginTop": 12,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1A1A1A",
              "fontFamily": "Nunito-Bold",
              "fontSize": 32,
              "lineHeight": 38,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        7
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1A1A1A",
              "fontFamily": "Nunito-Bold",
              "fontSize": 32,
              "lineHeight": 38,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        8
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1A1A1A",
              "fontFamily": "Nunito-Bold",
              "fontSize": 32,
              "lineHeight": 38,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        9
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "marginTop": 12,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1A1A1A",
              "fontFamily": "Nunito-Bold",
              "fontSize": 32,
              "lineHeight": 38,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        .
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#1A1A1A",
              "fontFamily": "Nunito-Bold",
              "fontSize": 32,
              "lineHeight": 38,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        0
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "height": 48,
          "justifyContent": "center",
          "opacity": 1,
          "width": 48,
        }
      }
    >
      <
        color="#1A1A1A"
        icon={
          {
            "icon": [
              576,
              512,
              [
                9003,
                "backspace",
              ],
              "f55a",
              "M576 128c0-35.3-28.7-64-64-64H205.3c-17 0-33.3 6.7-45.3 18.7L9.4 233.4c-6 6-9.4 14.1-9.4 22.6s3.4 16.6 9.4 22.6L160 429.3c12 12 28.3 18.7 45.3 18.7H512c35.3 0 64-28.7 64-64V128zM271 175c9.4-9.4 24.6-9.4 33.9 0l47 47 47-47c9.4-9.4 24.6-9.4 33.9 0s9.4 24.6 0 33.9l-47 47 47 47c9.4 9.4 9.4 24.6 0 33.9s-24.6 9.4-33.9 0l-47-47-47 47c-9.4 9.4-24.6 9.4-33.9 0s-9.4-24.6 0-33.9l47-47-47-47c-9.4-9.4-9.4-24.6 0-33.9z",
            ],
            "iconName": "delete-left",
            "prefix": "fas",
          }
        }
        size={24}
      />
    </View>
  </View>
</View>
`;
